google/cloud/pubsub/publisher_connection_test.cc

```diff
@@ -342,34 +347,71 @@ TEST(PublisherConnectionTest, HandleTransientEnabledRetry) {
 using ::google::cloud::testing_util::DisableTracing;
 using ::google::cloud::testing_util::EnableTracing;
 using ::google::cloud::testing_util::InstallSpanCatcher;
+using ::google::cloud::testing_util::OTelContextCaptured;
 using ::google::cloud::testing_util::SpanNamed;
+using ::google::cloud::testing_util::ThereIsAnActiveSpan;
 using ::testing::IsEmpty;
 using ::testing::UnorderedElementsAre;
 
 TEST(MakePublisherConnectionTest, TracingEnabled) {
   auto span_catcher = InstallSpanCatcher();
   auto mock = std::make_shared<pubsub_testing::MockPublisherStub>();
   Topic const topic("test-project", "test-topic");
+
+  auto mock_message_batch =
+      std::make_shared<pubsub_testing::MockMessageBatch>();
+  std::mutex mu;
+  std::condition_variable cv;
+  bool is_publish_complete = false;  // ABSL_GUARDED_BY(mu)
+  EXPECT_CALL(*mock_message_batch, SaveMessage(_)).Times(1);
+  EXPECT_CALL(*mock_message_batch, Flush).WillOnce([&] {
+    // Batch sink span should still be active when Flush is called.
+    EXPECT_TRUE(ThereIsAnActiveSpan());
+    EXPECT_TRUE(OTelContextCaptured());
+    return [&](auto) {
+      // Batch sink span should no longer be active, but there will be another
+      // active span so instead we can test the otel context is no longer
+      // captured.
+      EXPECT_FALSE(OTelContextCaptured());
+      // Wait until the lambda returned from Flush is executed to GetSpans.
+      {
+        std::lock_guard<std::mutex> lk(mu);
+        is_publish_complete = true;
+      }
+      cv.notify_all();
+    };
+  });
+
   EXPECT_CALL(*mock, AsyncPublish)
       .WillOnce([&](google::cloud::CompletionQueue&, auto,
                     google::pubsub::v1::PublishRequest const&) {
         google::pubsub::v1::PublishResponse response;
         response.add_message_ids("test-message-id-0");
         return make_ready_future(make_status_or(response));
       });
-  auto publisher =
-      MakeTestPublisherConnection(topic, mock, EnableTracing(Options{}));
+  auto publisher = MakeTestPublisherConnection(
+      topic, mock, EnableTracing(Options{}), mock_message_batch);
 
   auto response =
       publisher->Publish({MessageBuilder{}.SetData("test-data-0").Build()})
           .get();
+  publisher->Flush({});
+
+  // Wait until the BatchSink span ends before calling `GetSpans`.
+  std::unique_lock<std::mutex> lk(mu);
+  cv.wait(lk, [&] { return is_publish_complete; });
 
+  auto spans = span_catcher->GetSpans();
   EXPECT_THAT(
-      span_catcher->GetSpans(),
+      spans,
       UnorderedElementsAre(
           SpanNamed("projects/test-project/topics/test-topic send"),
           SpanNamed("publisher flow control"), SpanNamed("publish scheduler"),
-          SpanNamed("google.pubsub.v1.Publisher/Publish")));
+          SpanNamed("BatchSink::AsyncPublish"),
```

Comment:
When you do observe a flake: which span is missing?  What generates that span?  What ends the span?  What could cause it to not be ready at this point?

Reply:
"BatchSink::AsyncPublish" is always missing.

The generation is not the issue, it is that the span is not always "ended" before .GetSpans is called.

The span is ended in the callback returned from Flush. Before, I added a wait until the flush was called, but I should have added it to when the result of the flush.

Currently running it with 10,000 and it is passing up to 4000 times atm. So I think this is the correct fix.